pep517/build.py

```diff
@@ -3,25 +3,50 @@
 import argparse
 import logging
 import os
-import contextlib
 import pytoml
 import shutil
-import errno
-import tempfile
 
 from .envbuild import BuildEnvironment
 from .wrappers import Pep517HookCaller
+from .dirtools import tempdir, mkdir_p
 
 log = logging.getLogger(__name__)
 
 
-@contextlib.contextmanager
-def tempdir():
-    td = tempfile.mkdtemp()
+def validate_system(system):
+    """
+    Ensure build system has the requisite fields.
+    """
+    required = {'requires', 'build-backend'}
+    if required > set(system):
+        missing = required - set(system)
+        message = "Missing required fields: {missing}".format(**locals())
+        raise ValueError(message)
+
+
+def load_system(source_dir):
+    """
+    Load the build system from a source dir (pyproject.toml).
+    """
+    pyproject = os.path.join(source_dir, 'pyproject.toml')
+    with open(pyproject) as f:
+        pyproject_data = pytoml.load(f)
+    return pyproject_data['build-system']
+
+
+def compat_system(source_dir):
+    """
+    Given a source dir, attempt to get a build system backend
+    and requirements from pyproject.toml. Fallback to
+    setuptools.
+    """
     try:
-        yield td
-    finally:
-        shutil.rmtree(td)
+        system = load_system(source_dir)
+    except Exception:
```

Comment:
I do intend to trap both OSError and KeyError here... because the presence of a pyproject.toml without a `build-system` is comparable to no pyproject.toml (for this concern). I'd be tempted to say a malformed pyproject.toml should be treated as having no declared build system, though I do sympathize with your instinct here.

The problem I have with specifying `(OSError, KeyError)` in the `compat_system` function is it couples the internal implementation details of `load_system` with `compat_system`. I guess that's okay as one is meant to wrap the other, but it does make me a little uneasy.

Addressed in 0da9696.